File: solutions/haiku-1/haiku/go/anagrams.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strings"
)

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	wordGroups := make(map[string][]string)

	// Read words and group by sorted letters
	for scanner.Scan() {
		word := strings.TrimSpace(scanner.Text())
		if word == "" {
			continue
		}

		// Create a key from sorted letters
		sortedChars := strings.Split(word, "")
		sort.Strings(sortedChars)
		key := strings.Join(sortedChars, "")

		wordGroups[key] = append(wordGroups[key], word)
	}

	// Convert to slice for sorting
	type group struct {
		words []string
	}

	var groups []group
	for _, words := range wordGroups {
		sort.Strings(words)
		groups = append(groups, group{words: words})
	}

	// Sort groups by their first word
	sort.Slice(groups, func(i, j int) bool {
		return groups[i].words[0] < groups[j].words[0]
	})

	// Output
	for _, g := range groups {
		fmt.Println(strings.Join(g.words, " "))
	}
}

File: solutions/haiku-1/haiku/rust/anagrams.rs
use std::io::{self, BufRead};
use std::collections::BTreeMap;

fn main() {
    let stdin = io::stdin();
    let reader = stdin.lock();

    // Map from sorted letters to list of words
    let mut groups: BTreeMap<String, Vec<String>> = BTreeMap::new();

    for line in reader.lines() {
        if let Ok(word) = line {
            // Create the key by sorting the letters
            let mut sorted_chars: Vec<char> = word.chars().collect();
            sorted_chars.sort_unstable();
            let key: String = sorted_chars.iter().collect();

            // Add word to the group
            groups.entry(key).or_insert_with(Vec::new).push(word);
        }
    }

    // Sort words within each group and collect into a vector
    let mut sorted_groups: Vec<Vec<String>> = groups
        .into_values()
        .map(|mut group| {
            group.sort_unstable();
            group
        })
        .collect();

    // Sort groups by their first word
    sorted_groups.sort_unstable_by(|a, b| a[0].cmp(&b[0]));

    // Output
    for group in sorted_groups {
        println!("{}", group.join(" "));
    }
}
